src/renderer/atlas/AtlasEngine.api.cpp

```diff
@@ -0,0 +1,599 @@
+    SMALL_RECT rect;
+    rect.Top = gsl::narrow_cast<SHORT>(top);
+    rect.Bottom = gsl::narrow_cast<SHORT>(bottom);
+    return Invalidate(&rect);
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateSelection(const std::vector<SMALL_RECT>& rectangles) noexcept
+{
+    for (const auto& rect : rectangles)
+    {
+        // BeginPaint() protects against invalid out of bounds numbers.
+        // TODO: rect can contain invalid out of bounds coordinates when the selection is being
+        // dragged outside of the viewport (and the window begins scrolling automatically).
+        _api.invalidatedRows.x = gsl::narrow_cast<u16>(std::min<int>(_api.invalidatedRows.x, std::max<int>(0, rect.Top)));
+        _api.invalidatedRows.y = gsl::narrow_cast<u16>(std::max<int>(_api.invalidatedRows.y, std::max<int>(0, rect.Bottom)));
+    }
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateScroll(const COORD* const pcoordDelta) noexcept
+{
+    const auto delta = pcoordDelta->Y;
+    if (delta == 0)
+    {
+        return S_OK;
+    }
+
+    _api.scrollOffset = gsl::narrow_cast<i16>(clamp<int>(_api.scrollOffset + delta, i16min, i16max));
+
+    // InvalidateScroll() is a "synchronous" API. Any Invalidate()s after
+    // a InvalidateScroll() refer to the new viewport after the scroll.
+    // --> We need to shift the current invalidation rectangles as well.
+
+    _api.invalidatedCursorArea.top = gsl::narrow_cast<u16>(clamp<int>(_api.invalidatedCursorArea.top + delta, u16min, u16max));
+    _api.invalidatedCursorArea.bottom = gsl::narrow_cast<u16>(clamp<int>(_api.invalidatedCursorArea.bottom + delta, u16min, u16max));
+
+    if (delta < 0)
+    {
+        _api.invalidatedRows.x = gsl::narrow_cast<u16>(clamp<int>(_api.invalidatedRows.x + delta, u16min, u16max));
+        _api.invalidatedRows.y = _api.cellCount.y;
+    }
+    else
+    {
+        _api.invalidatedRows.x = 0;
+        _api.invalidatedRows.y = gsl::narrow_cast<u16>(clamp<int>(_api.invalidatedRows.y + delta, u16min, u16max));
+    }
+
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateAll() noexcept
+{
+    _api.invalidatedRows = invalidatedRowsAll;
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateCircling(_Out_ bool* const pForcePaint) noexcept
+{
+    RETURN_HR_IF_NULL(E_INVALIDARG, pForcePaint);
+    *pForcePaint = false;
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::InvalidateTitle(const std::wstring_view proposedTitle) noexcept
+{
+    WI_SetFlag(_api.invalidations, ApiInvalidations::Title);
+    return S_OK;
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateFont(const FontInfoDesired& fontInfoDesired, _Out_ FontInfo& fontInfo) noexcept
+{
+    return UpdateFont(fontInfoDesired, fontInfo, {}, {});
+}
+
+[[nodiscard]] HRESULT AtlasEngine::UpdateSoftFont(const gsl::span<const uint16_t> bitPattern, const SIZE cellSize, const size_t centeringHint) noexcept
+{
+    return S_OK;

```

Comment:
> return S_OK; [](http://example.com/codeflow?start=3&length=13)

lol more like E_NOTIMPL or S_FALSE

Reply:
That's true, but it's fine if this doesn't throw or log errors elsewhere for now. ^^

Reply:
yeah yeah.